packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -145,65 +171,426 @@ String getGradleVersionForAndroidPlugin(Directory directory, Logger logger) {
+
+  // Check if versions are newer than the max supported versions.
+  if (isWithinVersionRange(
+    javaV,
+    min: oneMajorVersionHigherJavaVersion,
+    max: '100.100',
+  )) {
+    // Assume versions Java versions newer than [maxSupportedJavaVersion]
+    // required a higher gradle version.
+    final bool validGradle = isWithinVersionRange(gradleV,
+        min: _maxKnownAndSupportedGradleVersion, max: '100.00');
+    logger.printWarning(
+        'Newer than known valid Java version ($javaV), gradle ($gradleV).'
+        '\n Treating as valid configuration.');
+    return validGradle;
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '3.4.0', max: '3.5.0')) {
-    return '5.6.2';
+
+  // Begin known Java <-> Gradle evaluation.
+    final List<JavaGradleCompat> compatList = <JavaGradleCompat>[
+    JavaGradleCompat(
+      javaMin: '19',
+      javaMax: '20',
+      gradleMin: '7.6',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '18',
+      javaMax: '19',
+      gradleMin: '7.5',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '17',
+      javaMax: '18',
+      gradleMin: '7.3',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '16',
+      javaMax: '17',
+      gradleMin: '7.0',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '15',
+      javaMax: '16',
+      gradleMin: '6.7',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '14',
+      javaMax: '15',
+      gradleMin: '6.3',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '13',
+      javaMax: '14',
+      gradleMin: '6.0',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '12',
+      javaMax: '13',
+      gradleMin: '5.4',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '11',
+      javaMax: '12',
+      gradleMin: '5.0',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    // 1.11 is a made up java version to cover everything in 1.10.*
+    JavaGradleCompat(
+      javaMin: '1.10',
+      javaMax: '1.11',
+      gradleMin: '4.7',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '1.9',
+      javaMax: '1.10',
+      gradleMin: '4.3',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '1.8',
+      javaMax: '1.9',
+      gradleMin: '2.0',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+  ];
+  for (final JavaGradleCompat data in compatList) {
+    if (isWithinVersionRange(javaV, min: data.javaMin, max: data.javaMax, inclusiveMax: false)) {
+      return isWithinVersionRange(gradleV, min: data.gradleMin, max: data.gradleMax);
```

Comment:
It looks like this class exists only for the sake of this list. So you could also skip having the class definition at all, and use the spiffy new Dart "records" feature:
```suggestion
    {javaMin: '1.8', javaMax: '1.9', gradleMin: '2.0'},
  ];
  for (final {:javaMin, :javaMax, :gradleMin} in compatList) {
    if (isWithinVersionRange(javaV, min: javaMin, max: javaMax, inclusiveMax: false)) {
      return isWithinVersionRange(gradleV, min: gradleMin, max: _maxKnownAndSupportedGradleVersion);
```

I think that would be quite clean, because it makes these records purely local to this loop — it avoids any worry about whether leaving out a `gradleMax` makes the class too specific for its name, or for hypothetical other ways one might use such a class.